### ISSUE TITLE
DatabaseObjectNotClosedException thrown on 9.8.0

### ISSUE THREAD

**Issue by betomaluje** [firebase/quickstart-android#176]
* Android device: Moto G (2nd Gen)
  * Android OS version: Android 6.0
  * Google Play Services version: 3.0.0
  * Firebase/Play Services SDK version: 9.8.0

Hi guys! I'm trying to use the latest version of the messaging library but I'm getting an `DatabaseObjectNotClosedException` exception and it says that the  `google_app_measurement_local.db` file is closed. The funny thing is that if I change the version of the library to, lets say 9.6.1, this doesn't happen and the notifications arrived successfully.

My gradle file is:
```
dependencies {
    compile fileTree(include: ['*.jar'], dir: 'libs')
    compile 'com.android.support:appcompat-v7:25.0.0'
    compile 'com.android.support:design:25.0.0'
    compile 'com.android.support:cardview-v7:25.0.0'
    compile 'com.google.android.gms:play-services-gcm:9.8.0'
    compile 'com.google.android.gms:play-services-maps:9.8.0'

    compile 'com.google.firebase:firebase-database:9.8.0'
    compile 'com.google.firebase:firebase-storage:9.8.0'
    compile 'com.google.firebase:firebase-auth:9.8.0'
    compile 'com.google.firebase:firebase-messaging:9.8.0'

    ... (other libraries)
}
```

(I'm trying to have support for both, GCM and FCM for now, just in case our FCM server breaks down we need a backup)

I've been looking on SO or similar and haven't found a solution to this.
Some links: 
- http://stackoverflow.com/questions/40364204/firebase-the-database-google-app-measurement-local-db-is-not-open
- http://stackoverflow.com/questions/40307855/firebase-is-firing-a-database-leak

So basically, my questions are:
- Why does this happen using the latest version and not the previous one?
- Is it feasible to have both, GCM and FCM in my app?

Thanks in advance

**Comment by samtstern** [comment]
@betomaluje sorry for the slow response here but we're looking into this!  Thank you for the report.

**Comment by betomaluje** [comment]
Hi! have you looked at this? I still can't upgrade the play services of my project

**Comment by gangarmitesh** [comment]
I am getting the same error and app crash, when it goes to background.

Got any solution?

**Comment by samtstern** [comment]
Hi all.  We have looked at this and there will be a fix in a future release.  Right now there are no known workarounds for the broken versions, but I will update here if there are.

**Comment by calvarez-ov** [comment]
Seeing this exception on 10.0.1, when enabling StrictMode.  No crash. Is there an issue opened on Firebase?

**Comment by sohailbud** [comment]
Seeing similar exception with 10.0.1

`12-30 13:07:42.527 513-522/com.sense.androidclient E/StrictMode: Finalizing a Cursor that has not been deactivated or closed. database = /data/user/0/com.sense.androidclient/databases/google_app_measurement_local.db, table = null, query = select count(1) from messages
                                                                 android.database.sqlite.DatabaseObjectNotClosedException: Application did not close the cursor or database object that was opened here
                                                                     at android.database.sqlite.SQLiteCursor.<init>(SQLiteCursor.java:98)
                                                                     at android.database.sqlite.SQLiteDirectCursorDriver.query(SQLiteDirectCursorDriver.java:50)
                                                                     at android.database.sqlite.SQLiteDatabase.rawQueryWithFactory(SQLiteDatabase.java:1318)
                                                                     at android.database.sqlite.SQLiteDatabase.rawQuery(SQLiteDatabase.java:1257)
                                                                     at com.google.android.gms.internal.zzatg.zza(Unknown Source)
                                                                     at com.google.android.gms.internal.zzatg.zza(Unknown Source)
                                                                     at com.google.android.gms.internal.zzatw.zzc(Unknown Source)
                                                                     at com.google.android.gms.internal.zzatu.zzb(Unknown Source)
                                                                     at com.google.android.gms.internal.zzatu.zza(Unknown Source)
                                                                     at com.google.android.gms.internal.zzatu$4.run(Unknown Source)
                                                                     at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428)
                                                                     at java.util.concurrent.FutureTask.run(FutureTask.java:237)
                                                                     at com.google.android.gms.internal.zzato$zzd.run(Unknown Source)`

**Comment by mattvb91** [comment]
Same problem here:

   compile 'com.google.firebase:firebase-core:9.8.0'
    compile 'com.google.firebase:firebase-crash:9.8.0'
    compile 'com.google.firebase:firebase-messaging:9.8.0'
    compile 'com.google.firebase:firebase-ads:9.8.0'

**Comment by JaredBanyard** [comment]
+1

**Comment by samtstern** [comment]
This issue is still known to be present in 10.0.1.

When a release goes out that contains the fix I will update this issue.

**Comment by mattvb91** [comment]
Thanks @samtstern. Any ETA on it? Unfortunately I have shipped a release with this and this only showed up in logs afterwards. It seems that this issue is causing the biggest problems on Android 6.0.1 phones for some reason?

**Comment by adamrhunter** [comment]
Seeing this on 6.0.1 with StrictMode -- for what it's worth: only occurs on first clean install.

**Comment by mhd-adeeb-masoud** [comment]
Still seeing this in 10.2.0 whenever an activity or fragment is destroyed.
`3-01 19:58:49.070 24205-24216/com.theclinician.oneclinic E/StrictMode: Finalizing a Cursor that has not been deactivated or closed. database = /data/user/0/com.theclinician.oneclinic/databases/google_app_measurement_local.db, table = null, query = select count(1) from messages
                                                                        android.database.sqlite.DatabaseObjectNotClosedException: Application did not close the cursor or database object that was opened here
                                                                            at android.database.sqlite.SQLiteCursor.<init>(SQLiteCursor.java:98)
                                                                            at android.database.sqlite.SQLiteDirectCursorDriver.query(SQLiteDirectCursorDriver.java:50)
                                                                            at android.database.sqlite.SQLiteDatabase.rawQueryWithFactory(SQLiteDatabase.java:1318)
                                                                            at android.database.sqlite.SQLiteDatabase.rawQuery(SQLiteDatabase.java:1257)
                                                                            at com.google.android.gms.internal.zzatv.zza(Unknown Source)
                                                                            at com.google.android.gms.internal.zzatv.zza(Unknown Source)
                                                                            at com.google.android.gms.internal.zzaul.zzc(Unknown Source)
                                                                            at com.google.android.gms.internal.zzauj.zzb(Unknown Source)
                                                                            at com.google.android.gms.internal.zzauj.zza(Unknown Source)
                                                                            at com.google.android.gms.internal.zzauj$8.run(Unknown Source)
                                                                            at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428)
                                                                            at java.util.concurrent.FutureTask.run(FutureTask.java:237)
                                                                            at com.google.android.gms.internal.zzaud$zzd.run(Unknown Source)
03-01 19:59:16.932 24205-24216/com.theclinician.oneclinic E/StrictMode: A resource was acquired at attached stack trace but never released. See java.io.Closeable for information on avoiding resource leaks.
                                                                        java.lang.Throwable: Explicit termination method 'close' not called
                                                                            at dalvik.system.CloseGuard.open(CloseGuard.java:180)
                                                                            at android.database.CursorWindow.<init>(CursorWindow.java:111)
                                                                            at android.database.AbstractWindowedCursor.clearOrCreateWindow(AbstractWindowedCursor.java:198)
                                                                            at android.database.sqlite.SQLiteCursor.fillWindow(SQLiteCursor.java:138)
                                                                            at android.database.sqlite.SQLiteCursor.getCount(SQLiteCursor.java:132)
                                                                            at android.database.AbstractCursor.moveToPosition(AbstractCursor.java:219)
                                                                            at android.database.AbstractCursor.moveToFirst(AbstractCursor.java:258)
                                                                            at com.google.android.gms.internal.zzatv.zza(Unknown Source)
                                                                            at com.google.android.gms.internal.zzatv.zza(Unknown Source)
                                                                            at com.google.android.gms.internal.zzaul.zzc(Unknown Source)
                                                                            at com.google.android.gms.internal.zzauj.zzb(Unknown Source)
                                                                            at com.google.android.gms.internal.zzauj.zza(Unknown Source)
                                                                            at com.google.android.gms.internal.zzauj$8.run(Unknown Source)
                                                                            at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428)
                                                                            at java.util.concurrent.FutureTask.run(FutureTask.java:237)
                                                                            at com.google.android.gms.internal.zzaud$zzd.run(Unknown Source)
03-01 19:59:16.933 24205-24216/com.theclinician.oneclinic E/StrictMode: Finalizing a Cursor that has not been deactivated or closed. database = /data/user/0/com.theclinician.oneclinic/databases/google_app_measurement_local.db, table = null, query = select count(1) from messages
                                                                        android.database.sqlite.DatabaseObjectNotClosedException: Application did not close the cursor or database object that was opened here
                                                                            at android.database.sqlite.SQLiteCursor.<init>(SQLiteCursor.java:98)
                                                                            at android.database.sqlite.SQLiteDirectCursorDriver.query(SQLiteDirectCursorDriver.java:50)
                                                                            at android.database.sqlite.SQLiteDatabase.rawQueryWithFactory(SQLiteDatabase.java:1318)
                                                                            at android.database.sqlite.SQLiteDatabase.rawQuery(SQLiteDatabase.java:1257)
                                                                            at com.google.android.gms.internal.zzatv.zza(Unknown Source)
                                                                            at com.google.android.gms.internal.zzatv.zza(Unknown Source)
                                                                            at com.google.android.gms.internal.zzaul.zzc(Unknown Source)
                                                                            at com.google.android.gms.internal.zzauj.zzb(Unknown Source)
                                                                            at com.google.android.gms.internal.zzauj.zza(Unknown Source)
                                                                            at com.google.android.gms.internal.zzauj$8.run(Unknown Source)
                                                                            at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428)
                                                                            at java.util.concurrent.FutureTask.run(FutureTask.java:237)
                                                                            at com.google.android.gms.internal.zzaud$zzd.run(Unknown Source)
03-01 19:59:23.901 24205-24216/com.theclinician.oneclinic E/StrictMode: A resource was acquired at attached stack trace but never released. See java.io.Closeable for information on avoiding resource leaks.
                                                                        java.lang.Throwable: Explicit termination method 'close' not called
                                                                            at dalvik.system.CloseGuard.open(CloseGuard.java:180)
                                                                            at android.database.CursorWindow.<init>(CursorWindow.java:111)
                                                                            at android.database.AbstractWindowedCursor.clearOrCreateWindow(AbstractWindowedCursor.java:198)
                                                                            at android.database.sqlite.SQLiteCursor.fillWindow(SQLiteCursor.java:138)
                                                                            at android.database.sqlite.SQLiteCursor.getCount(SQLiteCursor.java:132)
                                                                            at android.database.AbstractCursor.moveToPosition(AbstractCursor.java:219)
                                                                            at android.database.AbstractCursor.moveToFirst(AbstractCursor.java:258)
                                                                            at com.google.android.gms.internal.zzatv.zza(Unknown Source)
                                                                            at com.google.android.gms.internal.zzatv.zza(Unknown Source)
                                                                            at com.google.android.gms.internal.zzaul.zzc(Unknown Source)
                                                                            at com.google.android.gms.internal.zzauj.zzb(Unknown Source)
                                                                            at com.google.android.gms.internal.zzauj.zza(Unknown Source)
                                                                            at com.google.android.gms.internal.zzauj$8.run(Unknown Source)
                                                                            at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428)
                                                                            at java.util.concurrent.FutureTask.run(FutureTask.java:237)
                                                                            at com.google.android.gms.internal.zzaud$zzd.run(Unknown Source)
03-01 19:59:23.902 24205-24216/com.theclinician.oneclinic E/StrictMode: Finalizing a Cursor that has not been deactivated or closed. database = /data/user/0/com.theclinician.oneclinic/databases/google_app_measurement_local.db, table = null, query = select count(1) from messages
                                                                        android.database.sqlite.DatabaseObjectNotClosedException: Application did not close the cursor or database object that was opened here
                                                                            at android.database.sqlite.SQLiteCursor.<init>(SQLiteCursor.java:98)
                                                                            at android.database.sqlite.SQLiteDirectCursorDriver.query(SQLiteDirectCursorDriver.java:50)
                                                                            at android.database.sqlite.SQLiteDatabase.rawQueryWithFactory(SQLiteDatabase.java:1318)
                                                                            at android.database.sqlite.SQLiteDatabase.rawQuery(SQLiteDatabase.java:1257)
                                                                            at com.google.android.gms.internal.zzatv.zza(Unknown Source)
                                                                            at com.google.android.gms.internal.zzatv.zza(Unknown Source)
                                                                            at com.google.android.gms.internal.zzaul.zzc(Unknown Source)
                                                                            at com.google.android.gms.internal.zzauj.zzb(Unknown Source)
                                                                            at com.google.android.gms.internal.zzauj.zza(Unknown Source)
                                                                            at com.google.android.gms.internal.zzauj$8.run(Unknown Source)
                                                                            at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428)
                                                                            at java.util.concurrent.FutureTask.run(FutureTask.java:237)
                                                                            at com.google.android.gms.internal.zzaud$zzd.run(Unknown Source)
03-01 19:59:33.570 24205-24216/com.theclinician.oneclinic E/StrictMode: A resource was acquired at attached stack trace but never released. See java.io.Closeable for information on avoiding resource leaks.
                                                                        java.lang.Throwable: Explicit termination method 'close' not called
                                                                            at dalvik.system.CloseGuard.open(CloseGuard.java:180)
                                                                            at android.database.CursorWindow.<init>(CursorWindow.java:111)
                                                                            at android.database.AbstractWindowedCursor.clearOrCreateWindow(AbstractWindowedCursor.java:198)
                                                                            at android.database.sqlite.SQLiteCursor.fillWindow(SQLiteCursor.java:138)
                                                                            at android.database.sqlite.SQLiteCursor.getCount(SQLiteCursor.java:132)
                                                                            at android.database.AbstractCursor.moveToPosition(AbstractCursor.java:219)
                                                                            at android.database.AbstractCursor.moveToFirst(AbstractCursor.java:258)
                                                                            at com.google.android.gms.internal.zzatv.zza(Unknown Source)
                                                                            at com.google.android.gms.internal.zzatv.zza(Unknown Source)
                                                                            at com.google.android.gms.internal.zzaul.zzc(Unknown Source)
                                                                            at com.google.android.gms.internal.zzauj.zzb(Unknown Source)
                                                                            at com.google.android.gms.internal.zzauj.zza(Unknown Source)
                                                                            at com.google.android.gms.internal.zzauj$8.run(Unknown Source)
                                                                            at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428)
                                                                            at java.util.concurrent.FutureTask.run(FutureTask.java:237)
                                                                            at com.google.android.gms.internal.zzaud$zzd.run(Unknown Source)
03-01 19:59:33.570 24205-24216/com.theclinician.oneclinic E/StrictMode: Finalizing a Cursor that has not been deactivated or closed. database = /data/user/0/com.theclinician.oneclinic/databases/google_app_measurement_local.db, table = null, query = select count(1) from messages
                                                                        android.database.sqlite.DatabaseObjectNotClosedException: Application did not close the cursor or database object that was opened here
                                                                            at android.database.sqlite.SQLiteCursor.<init>(SQLiteCursor.java:98)
                                                                            at android.database.sqlite.SQLiteDirectCursorDriver.query(SQLiteDirectCursorDriver.java:50)
                                                                            at android.database.sqlite.SQLiteDatabase.rawQueryWithFactory(SQLiteDatabase.java:1318)
                                                                            at android.database.sqlite.SQLiteDatabase.rawQuery(SQLiteDatabase.java:1257)
                                                                            at com.google.android.gms.internal.zzatv.zza(Unknown Source)
                                                                            at com.google.android.gms.internal.zzatv.zza(Unknown Source)
                                                                            at com.google.android.gms.internal.zzaul.zzc(Unknown Source)
                                                                            at com.google.android.gms.internal.zzauj.zzb(Unknown Source)
                                                                            at com.google.android.gms.internal.zzauj.zza(Unknown Source)
                                                                            at com.google.android.gms.internal.zzauj$8.run(Unknown Source)
                                                                            at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428)
                                                                            at java.util.concurrent.FutureTask.run(FutureTask.java:237)
                                                                            at com.google.android.gms.internal.zzaud$zzd.run(Unknown Source)
`

**Comment by tsuijten** [comment]
Same here :(

**Comment by seviu** [comment]
It would be awesome if this gets fixed.

**Comment by coyarzun89** [comment]
+1

**Comment by AfzalivE** [comment]
Just close the poor cursor

**Comment by fuzzyBatman** [comment]
How frequently is this crashing? Did anyone by chance get a
'java.lang.RuntimeException: Could not read input channel file descriptors from parcel' crash?
because I got it once after adding firebase analytics.
Along with 'Finalizing a Cursor that has not been deactivated or closed' every time but no crash in StrictMode

**Comment by Malachiasz** [comment]
Same here:

`Finalizing a Cursor that has not been deactivated or closed. database = /data/user/0/com.mail.mobile.android.mail/databases/google_app_measurement_local.db, table = null, query = select count(1) from messages
 android.database.sqlite.DatabaseObjectNotClosedException: Application did not close the cursor or database object that was opened here
     at android.database.sqlite.SQLiteCursor.<init>(SQLiteCursor.java:98)
     at android.database.sqlite.SQLiteDirectCursorDriver.query(SQLiteDirectCursorDriver.java:50)
     at android.database.sqlite.SQLiteDatabase.rawQueryWithFactory(SQLiteDatabase.java:1318)
     at android.database.sqlite.SQLiteDatabase.rawQuery(SQLiteDatabase.java:1257)
     at com.google.android.gms.internal.zzatg.zza(Unknown Source)
     at com.google.android.gms.internal.zzatg.zza(Unknown Source)
     at com.google.android.gms.internal.zzatw.zzc(Unknown Source)
     at com.google.android.gms.internal.zzatu.zzb(Unknown Source)
     at com.google.android.gms.internal.zzatu.zza(Unknown Source)
     at com.google.android.gms.internal.zzatu$4.run(Unknown Source)
     at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428)
     at java.util.concurrent.FutureTask.run(FutureTask.java:237)
     at com.google.android.gms.internal.zzato$zzd.run(Unknown Source)`

**Comment by ricamgar** [comment]
Same here

**Comment by samtstern** [comment]
Can anyone provide a reproduction method for this?  Also does this crash the entire app or just appear in the logs?

**Comment by fuzzyBatman** [comment]
it happens when changing activities, or resuming one. Yes, the app crashes. It does not crash on run with StrictMode.setVmPolicy with penaltyDeath disabled though, it just keeps showing up in logcat

```
E/StrictMode: Finalizing a Cursor that has not been deactivated or closed. database = /data/user/0/< package name >/databases/google_app_measurement_local.db, table = null, query = select count(1) from messages
                                                                       android.database.sqlite.DatabaseObjectNotClosedException: Application did not close the cursor or database object that was opened here
                                                                           at android.database.sqlite.SQLiteCursor.<init>(SQLiteCursor.java:98)
                                                                           at android.database.sqlite.SQLiteDirectCursorDriver.query(SQLiteDirectCursorDriver.java:50)
                                                                           at android.database.sqlite.SQLiteDatabase.rawQueryWithFactory(SQLiteDatabase.java:1318)
                                                                           at android.database.sqlite.SQLiteDatabase.rawQuery(SQLiteDatabase.java:1257)
                                                                           at com.google.android.gms.internal.zzatv.zza(Unknown Source)
                                                                           at com.google.android.gms.internal.zzatv.zza(Unknown Source)
                                                                           at com.google.android.gms.internal.zzaul.zzc(Unknown Source)
                                                                           at com.google.android.gms.internal.zzauj.zzb(Unknown Source)
                                                                           at com.google.android.gms.internal.zzauj.zza(Unknown Source)
                                                                           at com.google.android.gms.internal.zzauj$8.run(Unknown Source)
                                                                           at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428)
                                                                           at java.util.concurrent.FutureTask.run(FutureTask.java:237)
                                                                           at com.google.android.gms.internal.zzaud$zzd.run(Unknown Source)
```

**Comment by samtstern** [comment]
@fuzzyBatman can you provide the code from your Activity `onCreate` or `onStart` that may be relevant?

**Comment by fuzzyBatman** [comment]
@samtstern this is the shortest onCreate() method in the project. onResume() and onStart() are empty

![create](https://cloud.githubusercontent.com/assets/5787026/24967080/9085d1b2-1fc6-11e7-970e-ea3728d753e3.PNG)

switching to this activity back and forth 3 times displayed the relevant log

```
D/Settings Activity: onCreate - entry
D/Settings Activity: onCreate - exit
D/Settings Activity: onStart
D/Settings Activity: onResume
D/OpenGLRenderer: endAllActiveAnimators on 0x8333d880 (RippleDrawable) with handle 0x7a7973d0
E/StrictMode: Finalizing a Cursor that has not been deactivated or closed. database = /data/user/0/<my-package-name>/databases/google_app_measurement_local.db, table = null, query = select count(1) from messages
              android.database.sqlite.DatabaseObjectNotClosedException: Application did not close the cursor or database object that was opened here
                  at android.database.sqlite.SQLiteCursor.<init>(SQLiteCursor.java:98)
                  at android.database.sqlite.SQLiteDirectCursorDriver.query(SQLiteDirectCursorDriver.java:50)
                  at android.database.sqlite.SQLiteDatabase.rawQueryWithFactory(SQLiteDatabase.java:1318)
                  at android.database.sqlite.SQLiteDatabase.rawQuery(SQLiteDatabase.java:1257)
                  at com.google.android.gms.internal.zzatv.zza(Unknown Source)
                  at com.google.android.gms.internal.zzatv.zza(Unknown Source)
                  at com.google.android.gms.internal.zzaul.zzc(Unknown Source)
                  at com.google.android.gms.internal.zzauj.zzb(Unknown Source)
                  at com.google.android.gms.internal.zzauj.zza(Unknown Source)
                  at com.google.android.gms.internal.zzauj$8.run(Unknown Source)
                  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428)
                  at java.util.concurrent.FutureTask.run(FutureTask.java:237)
                  at com.google.android.gms.internal.zzaud$zzd.run(Unknown Source)
E/StrictMode: A resource was acquired at attached stack trace but never released. See java.io.Closeable for information on avoiding resource leaks.
              java.lang.Throwable: Explicit termination method 'close' not called
                  at dalvik.system.CloseGuard.open(CloseGuard.java:180)
                  at android.database.CursorWindow.<init>(CursorWindow.java:111)
                  at android.database.AbstractWindowedCursor.clearOrCreateWindow(AbstractWindowedCursor.java:198)
                  at android.database.sqlite.SQLiteCursor.fillWindow(SQLiteCursor.java:138)
                  at android.database.sqlite.SQLiteCursor.getCount(SQLiteCursor.java:132)
                  at android.database.AbstractCursor.moveToPosition(AbstractCursor.java:219)
                  at android.database.AbstractCursor.moveToFirst(AbstractCursor.java:258)
                  at com.google.android.gms.internal.zzatv.zza(Unknown Source)
                  at com.google.android.gms.internal.zzatv.zza(Unknown Source)
                  at com.google.android.gms.internal.zzaul.zzc(Unknown Source)
                  at com.google.android.gms.internal.zzauj.zzb(Unknown Source)
                  at com.google.android.gms.internal.zzauj.zza(Unknown Source)
                  at com.google.android.gms.internal.zzauj$8.run(Unknown Source)
                  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428)
                  at java.util.concurrent.FutureTask.run(FutureTask.java:237)
                  at com.google.android.gms.internal.zzaud$zzd.run(Unknown Source)
```

**Comment by samtstern** [comment]
@fuzzyBatman thank you!  That is super helpful.

**Comment by samtstern** [comment]
Hey all.  I just heard back from the team and apparently the fix for this got delayed but will be included in the next SDK release, which should be in a matter of weeks.  I will keep this issue closed until it comes out, thank you all for your patience.

**Comment by hdtuan87** [comment]
`E/StrictMode: A resource was acquired at attached stack trace but never released. See java.io.Closeable for information on avoiding resource leaks.
                                                                      java.lang.Throwable: Explicit termination method 'close' not called
                                                                          at dalvik.system.CloseGuard.open(CloseGuard.java:180)
                                                                          at android.database.CursorWindow.<init>(CursorWindow.java:111)
                                                                          at android.database.AbstractWindowedCursor.clearOrCreateWindow(AbstractWindowedCursor.java:198)
                                                                          at android.database.sqlite.SQLiteCursor.fillWindow(SQLiteCursor.java:138)
                                                                          at android.database.sqlite.SQLiteCursor.getCount(SQLiteCursor.java:132)
                                                                          at android.database.AbstractCursor.moveToPosition(AbstractCursor.java:219)
                                                                          at android.database.AbstractCursor.moveToFirst(AbstractCursor.java:258)
                                                                          at com.google.android.gms.internal.zzatv.zza(Unknown Source)
                                                                          at com.google.android.gms.internal.zzatv.zza(Unknown Source)
                                                                          at com.google.android.gms.internal.zzaul.zzc(Unknown Source)
                                                                          at com.google.android.gms.internal.zzauj.zzb(Unknown Source)
                                                                          at com.google.android.gms.internal.zzauj.zza(Unknown Source)
                                                                          at com.google.android.gms.internal.zzauj$8.run(Unknown Source)
                                                                          at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428)
                                                                          at java.util.concurrent.FutureTask.run(FutureTask.java:237)
                                                                          at com.google.android.gms.internal.zzaud$zzd.run(Unknown Source)
05-04 10:04:58.039 16651-16663/com.bookmark.money.debug E/StrictMode: Finalizing a Cursor that has not been deactivated or closed. database = /data/user/0/com.bookmark.money.debug/databases/google_app_measurement_local.db, table = null, query = select count(1) from messages
                                                                      android.database.sqlite.DatabaseObjectNotClosedException: Application did not close the cursor or database object that was opened here
                                                                          at android.database.sqlite.SQLiteCursor.<init>(SQLiteCursor.java:98)
                                                                          at android.database.sqlite.SQLiteDirectCursorDriver.query(SQLiteDirectCursorDriver.java:50)
                                                                          at android.database.sqlite.SQLiteDatabase.rawQueryWithFactory(SQLiteDatabase.java:1318)
                                                                          at android.database.sqlite.SQLiteDatabase.rawQuery(SQLiteDatabase.java:1257)
                                                                          at com.google.android.gms.internal.zzatv.zza(Unknown Source)
                                                                          at com.google.android.gms.internal.zzatv.zza(Unknown Source)
                                                                          at com.google.android.gms.internal.zzaul.zzc(Unknown Source)
                                                                          at com.google.android.gms.internal.zzauj.zzb(Unknown Source)
                                                                          at com.google.android.gms.internal.zzauj.zza(Unknown Source)
                                                                          at com.google.android.gms.internal.zzauj$8.run(Unknown Source)
                                                                          at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428)
                                                                          at java.util.concurrent.FutureTask.run(FutureTask.java:237)
                                                                          at com.google.android.gms.internal.zzaud$zzd.run(Unknown Source)`

new update 10.2.4

**Comment by codesForLiving** [comment]
same issue on 10.2.4 . Is there an estimate, when it will be fixed?

**Comment by sandro-csimas** [comment]
Same issue on 10.2.6. When this issue will be fixed?

**Comment by petrovichtim** [comment]
+1 I got many crashes on Samsung GT-S5303B

**Comment by joshfriend** [comment]
@petrovichtim you should _never_ use strict mode in release builds

**Comment by petrovichtim** [comment]
@joshfriend  I don't use it. My stack trace is  `Exception android.database.CursorWindowAllocationException: Cursor window allocation of 2048 kb failed.
android.database.CursorWindow.<init> (CursorWindow.java:112)
android.database.CursorWindow.<init> (CursorWindow.java:100)
android.database.AbstractWindowedCursor.clearOrCreateWindow (AbstractWindowedCursor.java:198)
android.database.sqlite.SQLiteCursor.clearOrCreateWindow (SQLiteCursor.java:364)
android.database.sqlite.SQLiteCursor.fillWindow (SQLiteCursor.java:162)
android.database.sqlite.SQLiteCursor.getCount (SQLiteCursor.java:156)
android.database.AbstractCursor.moveToPosition (AbstractCursor.java:161)
android.database.AbstractCursor.move (AbstractCursor.java:197)
bre.b (bre.java:152)
bru.run (bru.java:3124)
java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1076)
java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:569)
java.lang.Thread.run (Thread.java:856)`

While I don't use SQLite in the project at all

**Comment by ABoschman** [comment]
It appears to be fixed in version 11.0.1.

**Comment by samtstern** [comment]
This should definitely be fixed in version `11.0.1`.  If anyone can reproduce on that version, please let me know.

**Comment by BoykoDmytro** [comment]
compile 'com.google.firebase:firebase-messaging:11.2.0'
Failed to open database '/data/user/0/xxx/databases/google_app_measurement_local.db'.
                                     android.database.sqlite.SQLiteCantOpenDatabaseException: unknown error android.database.sqlite.SQLiteCantOpenDatabaseException: unknown error (code 14): Could not open database
                                         at android.database.sqlite.SQLiteConnection.nativeOpen(Native Method)
                                         at android.database.sqlite.SQLiteConnection.open(SQLiteConnection.java:209)
                                         at android.database.sqlite.SQLiteConnection.open(SQLiteConnection.java:193)
                                         at android.database.sqlite.SQLiteConnectionPool.openConnectionLocked(SQLiteConnectionPool.java:463)
                                         at android.database.sqlite.SQLiteConnectionPool.open(SQLiteConnectionPool.java:185)
                                         at android.database.sqlite.SQLiteConnectionPool.open(SQLiteConnectionPool.java:177)
                                         at android.database.sqlite.SQLiteDatabase.openInner(SQLiteDatabase.java:808)
                                         at android.database.sqlite.SQLiteDatabase.open(SQLiteDatabase.java:793)
                                         at android.database.sqlite.SQLiteDatabase.openDatabase(SQLiteDatabase.java:696)
                                         at android.app.ContextImpl.openOrCreateDatabase(ContextImpl.java:690)
                                         at android.content.ContextWrapper.openOrCreateDatabase(ContextWrapper.java:299)
                                         at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked(SQLiteOpenHelper.java:223)
                                         at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase(SQLiteOpenHelper.java:163)
                                         at com.google.android.gms.internal.zzcbl.getWritableDatabase(Unknown Source)
                                         at com.google.android.gms.internal.zzcbk.getWritableDatabase(Unknown Source)
                                         at com.google.android.gms.internal.zzcbk.zzb(Unknown Source)
                                         at com.google.android.gms.internal.zzcbk.zza(Unknown Source)
                                         at com.google.android.gms.internal.zzceg.zzc(Unknown Source)
                                         at com.google.android.gms.internal.zzcdo.zzb(Unknown Source)
                                         at com.google.android.gms.internal.zzcdo.zza(Unknown Source)
                                         at com.google.android.gms.internal.zzcdw.run(Unknown Source)
                                         at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428)
                                         at java.util.concurrent.FutureTask.run(FutureTask.java:237)
                                         at com.google.android.gms.internal.zzccn.run(Unknown Source)

**Comment by himangipatel** [comment]
**I am using latest version with 11.8.0
still getting the same error**


SQLiteStatements: /data/user/0/com.test/databases/google_app_measurement_local.db: "COMMIT;"
01-26 14:58:31.212 27111-27146/com.test V/SQLiteTime: /data/user/0/com.test/databases/google_app_measurement_local.db: "COMMIT;" took 4.000 ms
01-26 14:58:31.213 27111-27146/com.test V/SQLiteCursor: Cursor close object=160000322
01-26 14:58:31.216 27111-27146/com.test D/SQLiteDatabase: dispose database false
                                                                    java.lang.Throwable: stacktrace
                                                                        at android.database.sqlite.SQLiteDatabase.dispose(SQLiteDatabase.java:286)
                                                                        at android.database.sqlite.SQLiteDatabase.onAllReferencesReleased(SQLiteDatabase.java:280)
                                                                        at android.database.sqlite.SQLiteClosable.releaseReference(SQLiteClosable.java:74)
                                                                        at android.database.sqlite.SQLiteClosable.close(SQLiteClosable.java:106)
                                                                        at com.google.android.gms.internal.zzchi.zzeb(Unknown Source)
                                                                        at com.google.android.gms.internal.zzckg.zza(Unknown Source)
                                                                        at com.google.android.gms.internal.zzcko.run(Unknown Source)
                                                                        at com.google.android.gms.internal.zzckg.zzj(Unknown Source)
                                                                        at com.google.android.gms.internal.zzckg.zzc(Unknown Source)
                                                                        at com.google.android.gms.internal.zzcjn.zzb(Unknown Source)
                                                                        at com.google.android.gms.internal.zzcjn.zza(Unknown Source)
                                                                        at com.google.android.gms.internal.zzcjv.run(Unknown Source)
                                                                        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:423)
                                                                        at java.util.concurrent.FutureTask.run(FutureTask.java:237)
                                                                        at com.google.android.gms.internal.zzcil.run(Unknown Source)
